streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java

```diff
@@ -0,0 +1,476 @@
+     */
+    public final static class Sink extends AbstractNode {
+        private final String topic;
+
+        Sink(final String name,
+             final String topic) {
+            super(name);
+            this.topic = topic;
+        }
+
+        /**
+         * The topic name this sink node is writing to.
+         * @return a topic name
+         */
+        public String topic() {
+            return topic;
+        }
+
+        @Override
+        void addSuccessor(final Node successor) {
+            throw new UnsupportedOperationException("Sinks don't have successors.");
+        }
+
+        @Override
+        public String toString() {
+            return "Sink: " + name + "(topic: " + topic + ") <-- " + nodeNames(predecessors);
+        }
+
+        @Override
+        public boolean equals(final Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            final Sink sink = (Sink) o;
+            return name.equals(sink.name)
+                && topic.equals(sink.topic)
+                && predecessors.equals(sink.predecessors);
+        }
+
+        @Override
+        public int hashCode() {
+            // omit predecessors as it might change and alter the hash code
+            return Objects.hash(name, topic);
+        }
+    }
+
+    void addSubtopology(final Subtopology subtopology) {
+        subtopologies.add(subtopology);
+    }
+
+    void addGlobalStore(final GlobalStore globalStore) {
+        globalStores.add(globalStore);
+    }
+
+    /**
+     * All sub-topologies of the represented topology.
+     * @return set of all sub-topologies
+     */
+    public Set<Subtopology> subtopologies() {
+        return Collections.unmodifiableSet(subtopologies);
+    }
+
+    /**
+     * All global stores of the represented topology.
+     * @return set of all global stores
+     */
+    public Set<GlobalStore> globalStores() {
+        return Collections.unmodifiableSet(globalStores);
+    }
+
+    @Override
+    public String toString() {
+        return subtopologiesAsString() + globalStoresAsString();
+    }
+
+    private static String nodeNames(final Set<Node> nodes) {
+        final StringBuilder sb = new StringBuilder();
+        if (!nodes.isEmpty()) {
+            for (final Node n : nodes) {
+                sb.append(n.name());
+                sb.append(", ");
+            }
+            sb.deleteCharAt(sb.length() - 1);
+            sb.deleteCharAt(sb.length() - 1);
+        }
+        return sb.toString();
+    }
+
+    private String subtopologiesAsString() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append("Sub-topologies: \n");
+        if (subtopologies.isEmpty()) {
+            sb.append("  none\n");
+        } else {
+            for (final Subtopology st : subtopologies) {
+                sb.append("  ");
```

Comment:
The indention is basically hard coded and this renders fine. We don't render the structure of sub-topologies but only list all nodes within it. I did print `GlobalKTableIntegartionTest#shouldKStreamGlobalKTableLeftJoin`:
```
Sub-topologies: 
  Sub-topology: 2
    Processor: KTABLE-SOURCE-0000000004(stores: [table]) -->  <-- KSTREAM-SOURCE-0000000003
    Source: KSTREAM-SOURCE-0000000003(topics: input-table-1) --> KTABLE-SOURCE-0000000004
  Sub-topology: 1
    Processor: KSTREAM-LEFTJOIN-0000000005(stores: []) --> KSTREAM-FOREACH-0000000006 <-- KSTREAM-SOURCE-0000000002
    Processor: KSTREAM-FOREACH-0000000006(stores: []) -->  <-- KSTREAM-LEFTJOIN-0000000005
    Source: KSTREAM-SOURCE-0000000002(topics: input-stream-1) --> KSTREAM-LEFTJOIN-0000000005
Global Stores:
  GlobalStore: KSTREAM-SOURCE-0000000000(topic: globalOne-1) -> KTABLE-SOURCE-0000000001(store: globalStore)
```

The "arrows" indicate predecessor and successors for each node. It's not quite pretty and I am happy to improve this if you have good idea. However, the main purpose of this API is not to just print the topology to stdout but you use if for own rendering. Also, anybody can to a stdout pretty print too by implementing their own traversal strategy that produces a `String`.